### ISSUE TITLE
Unit test failures (DL4J, ND4J, DataVec)

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#2784]
Current failures:

**ND4J, native**
- [ ] RngTests.testRandomWithOrder:64 expected:<0.778587281703949> but was:<0.0>
- [x] ~~TestTensorAlongDimension.testTadShapes2d:157 getAll3dTestArraysWithShape(12345,[3, 4, 5]).get(0)/Nd4j.linspace(1,len,len)([3, 4, 5],'c') 1 expected:<[2,3,4,20,5,1,-1,99]> but was:<[2,3,4,20,5,0,-1,99]>~~
- [x] ~~OpExecutionerTestsC.testEuclideanManhattanDistanceAlongDimension_Rank4:804 expected:<[3,2,2,2,4,2,1,8,-1,99]> but was:<[3,2,2,2,4,2,1,0,-1,99]>~~
- [ ] OperationProfilerTests.testBadTad4:195
- [ ] OperationProfilerTests.testBlasFF:253 » Arithmetic / by zero

**ND4J, cuda**
- [x] ~~Crash: https://gist.github.com/AlexDBlack/c21fe1c9dc565df8d00ec8bb37cdff1e~~

**DataVec**
- [x] ~~TestCustomTransformJsonYaml.testCustomTransform:53 » Runtime org.nd4j.shade.ja...~~ Fixed here: https://github.com/deeplearning4j/DataVec/pull/180

**DL4J Unit Tests - Native**
- [x] ~~org.deeplearning4j.datasets.iterator.DataSetIteratorTest.testBatchSizeOfOne(DataSetIteratorTest.java:77)~~ (split used in example doesn't guarantee exact output size)
- [x] ~~org.deeplearning4j.nn.updater.TestUpdaters.testMultiLayerUpdater(TestUpdaters.java:444): IllegalStateException: Mis matched shapes~~
- [x] ~~org.deeplearning4j.nn.graph.TestVariableLengthTSCG.testInputMasking: IllegalStateException: Mismatched shapes (shape = [5, 2], row vector shape =[1, 5])~~
- [x] org.deeplearning4j.util.ModelGuesserTest.testModelGuess(ModelGuesserTest.java:30) Unknown keras weight initializer ~~XAVIER~~ "orthogonal". (fails due to a legitimate reason; error message was also wrong but now fixed)
- [x] ~~org.deeplearning4j.nn.modelimport.keras.KerasLayerTest.testBuildSubsamplingLayer: expected: org.deeplearning4j.nn.conf.layers.SubsamplingLayer$PoolingType<MAX> but was: org.deeplearning4j.nn.conf.layers.PoolingType<MAX>~~
Fixes: https://github.com/deeplearning4j/deeplearning4j/pull/2785

**DL4J Unit Tests - CUDA**
- [x] RBMTests.testActivate
- [ ] TestReconstructionDistributions.gradientCheckReconsturctionDistributions (seems like numerical precision issue)
- [x] Possible infinite loop on TestDeepWalk.testParallel (this is the old implementation)
- [x] testEarlyStoppingSpark (multiple failures)
- [x] TestSparkComputationGraph.testSeedRepeatability
- [x] TestCompareParameterAveragingSparkVsSingleMachine - multiple failures (NaNs in params)
- [x] TestSparkMultiLayerParameterAveraging.testFixViaStringPathsComGraph  (NaNs in params)
(Otherwise: haven't run CUDA NLP tests yet; testModelGuess also failing as per native)

**Comment by agibsonccc** [comment]
https://github.com/deeplearning4j/nd4j/pull/1616 @AlexDBlack This is for TestTensorAlongDimension.

**Comment by raver119** [comment]
libnd4j cuda was updated to reflect all TAD changes, all nd4j-tests pass now.

**Comment by raver119** [comment]
btw, all tests pass for me on native as well. @AlexDBlack please poke me if you still see them failing.

**Comment by huitseeker** [comment]
Edit: for **DL4J** at 59fa51faedbd06fd3dc1c19e0b6234984b750093, the only tests I see failing are:
- **Native & CUDA**
  ModelGuesserTest.testModelGuess:30 » UnsupportedKerasConfiguration Unknown ker...
  (see #2745)
- **CUDA**
  TestReconstructionDistributions.gradientCheckReconstructionDistributions:359 expected:<0> but was:<1>

**Comment by AlexDBlack** [comment]
I've run through dl4j core tests on native and (most) tests on cuda also before merging this: https://github.com/deeplearning4j/deeplearning4j/pull/2872

@huitseeker I'm not able to reproduce those gradient check failures anymore with CUDA (I had been able to, as of the time of opening this issue). If you still see it on master + cuda, open another issue. Otherwise I'll be re-running everything in preparation for release anyway.

Still seeing some nd4j issues, however: I've opened a separate issue for that here: https://github.com/deeplearning4j/nd4j/issues/1655
